### PR TITLE
Add alert for blackbox exporter endpoint down

### PR DESCRIPTION
**Pull request by tfussell**
Towards https://github.com/giantswarm/telekom-panamax/issues/104

This PR:

- Adds an alerting rule for blackbox exporter endpoint down on KVM.

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [x] Request review from oncall area, as well as team (e.g: `oncall-kaas-cloud` GitHub group).

### CODE REVIEW

File: helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml [hunk 1]
```diff
@@ -12,6 +12,17 @@ spec:
   groups:
   - name: kvm
     rules:
+    - alert: BlackboxExporterEndpointDown
+      annotations:
+        description: '{{`Blackbox exporter endpoint ({{ $labels.instance }}) is down.`}}'
+        opsrecipe: blackbox-exporter-endpoint-down/
```
Comment (tfussell) [inline helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:18]: I'll add this Ops Recipe while I wait for review.

Comment (tfussell) [reply helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:18]: Recipe here https://github.com/giantswarm/prometheus-rules/pull/59

**Review by MarcelMue — APPROVED** [review]

File: helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml (hunk 1 above)
Comment (MarcelMue) [reply helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:18]: The link to the recipe is this PR and not the recipe PR :D

**Comment by wind0r** [comment]
15 minutes of one probe being down will "never" report any of the issues we currently have with kvm.

A Screenshot of the outage/issue from today shows quite good that this rule will not help to spot those problems
image

I think something like `if $x request in 5 minutes are over $y ms` would be better.
Avg blackbox response time is 4ms

**Comment by wind0r** [comment]
@tfussell Just, I would have expected Giant Swarm to handle this. Some nginx pods didn't accept any traffic. Some info here: https://github.com/giantswarm/telekom-panamax/issues/106

This kinda was the first or second time nginx issues where spotted via Blackbox Exporter. Most of the time we spot broken overlay network.

The good part with those blackbox exporters is that "just" Giant Swarm stuff is involved. It can't be an application level issue on our side, since they just check if nginx is accessible.
If Loadbalancers cant access the cluster or responses are slow in my opinion it is something Giant Swarm should handle.

File: helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml [hunk 2]
```diff
@@ -12,6 +12,28 @@ spec:
...
+      expr: up{job="management-cluster-blackbox-ingress-http"} == 0
+      for: 15m
+      labels:
+        area: kaas
+        severity: page
+        team: rocket
+        topic: observability
+    - alert: BlackboxExporterEndpointNotResponding
```
Comment (tfussell) [inline helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:26]: Added a new alert that actually uses the metrics reported by the blackbox exporter.

**Comment by tfussell** [comment]
> @tfussell Just, I would have expected Giant Swarm to handle this. Some nginx pods didn't accept any traffic. Some info here: [giantswarm/telekom-panamax#106](https://github.com/giantswarm/telekom-panamax/issues/106)

I see. Thank you for the additional info. I added a new alert which should cover this use case. It is not configured to page us currently as we need to determine the correct alerting thresholds. We will check how often it is firing next week and make it paging once we have a good sense of what the threshold should be.

File: helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml [hunk 3]
```diff
@@ -12,6 +12,28 @@ spec:
...
+      annotations:
+        description: '{{`Blackbox exporter endpoint ({{ $labels.instance }}) is not responding.`}}'
+        opsrecipe: blackbox-exporter-endpoint-down/
+      expr: sum by(instance) (avg_over_time(probe_http_duration_seconds[5m])) > 0.2
```
Comment (pipo02mix) [inline helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:30]: Out of curiosity, why 0.2 secs?

**Review by pipo02mix — APPROVED** [review]
it seems fair

File: helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml (hunk 3 above)
Comment (tfussell) [reply helm/prometheus-rules/templates/alerting-rules/kvm.management-cluster.rules.yml:30]: It was chosen arbitrarily based on the last few days of metrics. I wanted it to be high enough that normal variation didn't cause it to fire, but real issues would. I figure we can adjust it downwards if we miss a problem we should have been paged for or upwards if we're getting spurious alerts.

**Comment by tfussell** [comment]
I will merge tomorrow morning just in case there are spurious pages.

**Comment by webwurst** [comment]
@wind0r Can you point us to more incidents to test our query? For now I have `2021-08-13 10:00 - 16:00` for example.

**Comment by wind0r** [comment]
@webwurst the other 2 nginx incidents are over 30 days old so data is gone :( sorry

**Comment by ubergesundheit** [comment]
@tfussell any progress regarding the alert query?

**Comment by tfussell** [comment]
> @tfussell any progress regarding the alert query?

I now have a rule which will reliably indicate the August 13th issue based on the historical data. I have asked for help from SIG Monitoring making sure it is performant. Tomorrow I will try to reproduce the flannel downtime to ensure it would catch that problem as well.

**Comment by tfussell** [comment]
Closing in favor of https://github.com/giantswarm/prometheus-rules/pull/94

**Pull request closed without merging**